keps/sig-multicluster/5339-clusterprofile-plugin-credentials/README.md

```diff
@@ -0,0 +1,619 @@
+```
+type Cluster struct {
+  // Server is the address of the kubernetes cluster (https://hostname:port).
+  Server string `json:"server"`
+  // TLSServerName is passed to the server for SNI and is used in the client to
+  // check server certificates against. If ServerName is empty, the hostname
+  // used to contact the server is used.
+  // +optional
+  TLSServerName string `json:"tls-server-name,omitempty"`
+  // InsecureSkipTLSVerify skips the validity check for the server's certificate.
+  // This will make your HTTPS connections insecure.
+  // +optional
+  InsecureSkipTLSVerify bool `json:"insecure-skip-tls-verify,omitempty"`
+  // CAData contains PEM-encoded certificate authority certificates.
+  // If empty, system roots should be used.
+  // +listType=atomic
+  // +optional
+  CertificateAuthorityData []byte `json:"certificate-authority-data,omitempty"`
+  // ProxyURL is the URL to the proxy to be used for all requests to this
+  // cluster.
+  // +optional
+  ProxyURL string `json:"proxy-url,omitempty"`
+  // Config holds additional config data that is specific to the exec
+  // plugin with regards to the cluster being authenticated to.
+  //
+  // This data is sourced from the clientcmd Cluster object's
+  // extensions[client.authentication.k8s.io/exec] field:
+  //
+  // clusters:
+  // - name: my-cluster
+  //   cluster:
+  //     ...
+  //     extensions:
+  //     - name: client.authentication.k8s.io/exec  # reserved extension name for per cluster exec config
+  //       extension:
+  //         audience: 06e3fbd18de8  # arbitrary config
+  //
+  // In some environments, the user config may be exactly the same across many clusters
+  // (i.e. call this exec plugin) minus some details that are specific to each cluster
+  // such as the audience.  This field allows the per cluster config to be directly
+  // specified with the cluster info.  Using this field to store secret data is not
+  // recommended as one of the prime benefits of exec plugins is that no secrets need
+  // to be stored directly in the kubeconfig.
+  // +optional
+  Config runtime.RawExtension `json:"config,omitempty"`
+}
+```
+
+This data must be created from the ClusterProfile, as described in the following section.
+
+### Standardizing the Endpoint Property
+
+In order to populate the Cluster object that the exec provider requires, we standardize the following properties.
+
+Property: `multicluster.sigs.io/execprovider-cluster`
+Value: json containing the Cluster struct (see above).
+Example Value: "{\"Server\":\"https://example.com:443\"}"
+
+### Configuring plugins in the controller
+
+Plugins are designated by a hardcoded string, for example, "gke" for GKE Clusters, which allows the controller
+to attach a different binary name or path for the the binary. It is expected that the library will have a mapping from
+its supported plugin names to the expected binary to call. It is expected that the mapping is provided via a flag on the controller.
```

Comment:
Wouldn't mapping binaries make this vendor/implementation specific? It's probably not useful to just hardcode the binary if you don't know how it's configured, right? I.e. hardcoding a binary only specifies the `Command` part of the [ `ExecConfig`](https://github.com/kubernetes/kubernetes/blob/a34c07971b610eb33908a743eadb4c61beeecc50/staging/src/k8s.io/client-go/tools/clientcmd/api/types.go#L206-L208).

Reply:
I added clarification between the string and the plugin.

The string to map to plugin is the "credentials type", then the controller has to setup a plugin for that "credentials type".

A credentials Type could be: `google`, `azure`, `direct` and then the controller would have to plug the right plugin (there may be multiple that works for a given credentials).

For google, I'm guessing one could use workload identity and therefore use a plugin called `google-with-wi` which would read local files/env to use WI that is a google creds. That same, google identity could map to a `google-user` which would use an interactive login?